app/src/main/java/com/github/orkest/Model/AudioRecognition.kt

```diff
@@ -0,0 +1,174 @@
+        fun recordingFlow(coroutineScope: CoroutineScope): Flow<AudioChunk> = flow {
+            while (currentCoroutineContext().isActive) {
+                emit(audioChunkRecord())
+                delay(100)
+            }
+        }
+
+
+        //TODO: add dependency injection for the recordingFlow
+        @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+        suspend fun recognizeSong(recordingFlow: Flow<AudioChunk>,
+                                  coroutineScope: CoroutineScope): CompletableFuture<String>{
+
+            val currentSession = (ShazamKit.createStreamingSession(
+                catalog,
+                AudioSampleRateInHz.SAMPLE_RATE_48000,
+                Constants.SHAZAM_SESSION_READ_BUFFER_SIZE
+            ) as ShazamKitResult.Success).data
+
+
+            // val coroutineScope = CoroutineScope(Dispatchers.Default)
+
+            // records audio and flows it to the StreamingSession
+            coroutineScope.launch {
+                recordingFlow.collect { audioChunk ->
+                    currentSession.matchStream(
+                        audioChunk.buffer,
+                        audioChunk.meaningfulLengthInBytes,
+                        audioChunk.timestamp
+                    )
+                }
+            }
+
+            val title = CompletableFuture<String>()
+            // collect the results
+            coroutineScope.launch {
+                currentSession.recognitionResults().collect { matchResult ->
+                    if (matchResult is MatchResult.Match) {
+                        title.complete(matchResult.matchedMediaItems[0].title!!)
+                    }
+                    if (matchResult is MatchResult.NoMatch) {
+                        title.complete("No match found")
+                    }
+
+                    if (matchResult is MatchResult.Error) {
+                        title.completeExceptionally(matchResult.exception)
+                    }
+                }
+            }
+            return title
+        }
+    }
+
+    data class AudioChunk (
+        val buffer: ByteArray = ByteArray(Constants.SHAZAM_SESSION_READ_BUFFER_SIZE),
+        val meaningfulLengthInBytes: Int = Constants.SHAZAM_SESSION_READ_BUFFER_SIZE,
+        val timestamp: Long = System.currentTimeMillis()
+    ) {
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (javaClass != other?.javaClass) return false
+
+            other as AudioChunk
+
+            return buffer.contentEquals(other.buffer) &&
+                    meaningfulLengthInBytes == other.meaningfulLengthInBytes &&
+                    timestamp == other.timestamp
+        }
+
+        override fun hashCode(): Int {
+            var result = buffer.contentHashCode()
+            result = 31 * result + meaningfulLengthInBytes
+            result = 31 * result + timestamp.hashCode()
+            return result
+        }
+    }
+}
```

Comment:
Very good, but a bit hard to understand without comment

Reply:
Yes sorry for that it will be added in next sprint